linera-service/src/linera-exporter/state.rs

```diff
@@ -1,84 +1,195 @@
 // Copyright (c) Zefchain Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
+use std::{
+    marker::PhantomData,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+
 use linera_base::{crypto::CryptoHash, data_types::BlockHeight, identifiers::ChainId};
 use linera_client::config::DestinationId;
-use linera_sdk::views::{RootView, View, ViewError};
+use linera_sdk::{
+    ensure,
+    views::{RootView, View},
+};
 use linera_views::{
-    context::Context, map_view::MapView, reentrant_collection_view::ReentrantCollectionView,
-    register_view::RegisterView, views::ClonableView,
+    context::Context, log_view::LogView, map_view::MapView, register_view::RegisterView,
+    views::ClonableView,
 };
+use serde::{de::Visitor, ser::SerializeSeq as _, Deserialize, Serialize};
 
-use crate::ExporterError;
+use crate::common::{BlockId, ExporterError, LiteBlockId};
 
 /// State of the linera exporter as a view.
 #[derive(Debug, RootView, ClonableView)]
 pub struct BlockExporterStateView<C> {
+    /// The causal state.
+    canonical_state: LogView<C, CryptoHash>,
```

Comment:
Why write `causal state` and name it `canonical_state`.

Also, the `CanonicalState` type in `storage.rs`. Is it related?